mautic/mautic
# Optimizing contacts activity API (refactoring of PR-10237 for Mautic v5)

This is a refactoring of [PR-10237](https://github.com/mautic/mautic/pull/10237) for Mautic v5 as [requested](https://github.com/mautic/mautic/pull/12187#issuecomment-1504948090). See original PR for feature and testing description.

Info: This refactoring covers 100% of the original PR, except 1 file which have been skipped because it exists already in 5.x branch: `app/bundles/CoreBundle/Doctrine/PreUpAssertionMigration.php`.

Fixes https://github.com/mautic/mautic/issues/12146

## Description

Add indicies to columns ( especially those that use file sort) plus other fixes:

- Set the event type name to default in the beginning of each for loop (PageBundle/EventListener/LeadSubscriber::onTimelineGenerate)
- Use second parameter for ordering in TimelineTrait
- Remove 'where stat.lead_id = null' from EmailBundle/Entity/EmailReplyRepository::getByLeadForTimeline
- USE INDEX in CampaignBundle/Entity/LeadEventLogRepository::getLeadLogs

## Steps to test this PR

1. Load up [this PR](https://gitpod.io/#https://github.com/mautic/mautic/pull/12305)
2. Test that the API for all affected parts work as before (https://developer.mautic.org/#rest-api), mainly:

/api/contacts/2/activity
/api/contacts/activity

3. Check the migrations work (if possible!)
